### ISSUE TITLE
Having issues with server visibility and server querys

### ISSUE THREAD

**Issue by Freilichtbuehne** [Facepunch/garrysmod-issues#5116]
### Details

Since the latest december update our server is facing following issues:
- Gameserver Viewer in our forum (based on [source query protocol](https://developer.valvesoftware.com/wiki/Server_queries) and the request A2S_INFO) not working
       - Header of responce returns `0x65` instead of `0x49` and I have no clue why
- Players have trouble finding server in server list (not for everybody)
- Server disappeared in favorites
- 90 % of other servers are flagged "Timeout" in monitoring tool "HLSW" (same reason as first point)
- Users connecting via steam connect link having following error: 
      - https://i.ibb.co/DLX640Q/image.png

I know that this isn't directly a bug. But since the latest update, player have trouble finding our server.

I have not found any change that could be attributed to the problem.

### Steps to reproduce
None. Permanent.

**Comment by MaxTM1** [comment]
I have the same problem....

**Comment by Luke-Ufo** [comment]
I tried to fix it for 3 hours but didn't succeed.

**Comment by robotboy655** [comment]
Tell those users to update their Steam clients. Contact your server provider for any issues with firewall rules.

See this official post from Valve:
https://steamcommunity.com/discussions/forum/14/2989789048633291344/

**Comment by matiasg90** [comment]
Since the last update, no gmod server on windows is shown in HLSW

**Comment by robotboy655** [comment]
3rd party tools will need to be updated, they had over a year to do it.

**Comment by ThatsLucas** [comment]
same issue

**Comment by paulfuriox** [comment]
same issue ....

**Comment by ExxtazzTV** [comment]
same problem for my server some people see it and many others don't

**Comment by MauriceKrabs** [comment]
same problem since update, loss of 50% of players

**Comment by michak89** [comment]
Same issue, sometimes people can see it and sometimes cannot, after few minutes it is visible again + over 50% less players.

**Comment by jorjic** [comment]
> Header of responce returns 0x65 instead of 0x49 and I have no clue why

The query protocol was changed and now some packets require challenge verification to prevent abuse. The `0x65` is `A` in ASCII and [is defined here](https://github.com/VSES/SourceEngine2007/blob/master/src_main/common/proto_oob.h#L95) for `S2C_CHALLENGE`. You should [inform the author of the plugin](https://www.woltlab.com/user/1494075-titusk/) that the protocol has updated and they need to update immediately.


@robotboy655 With regards to the official Valve post, FletcherDunn said:
> We will post again when the vast majority of users are running clients that understand the new protocol, and enabling the new protocol is safe.

Did they ever announce that enabling the new protocol is safe? Comments on the discussion thread indicate that the changes were activated and made default to Team Fortress 2 in September without a word from FletcherDunn.

Also, I understand the changes to the protocol involved hardening challenge authentication and adding challenge verification to A2S_INFO packets. Can you confirm if Garry's Mod still uses Fortran's RAN1 pseudorandom number generator for challenge generation? [This issue](https://github.com/danielga/gmsv_serversecure/issues/72) in another repo indicates that this was being exploited in the wild since September.

As for the issue people seem to be having here, we desperately need a packet capture on either end of the problem to tell what's *actually going on*.

**Comment by robotboy655** [comment]
> Did they ever announce that enabling the new protocol is safe?

Not that I am aware of. Apparently this behavior was force enabled within Steamworks API.

> Can you confirm if Garry's Mod still uses Fortran's RAN1 pseudorandom number generator for challenge generation?

No as the entire system is inside Steamworks binaries, we have no control over it. (besides just recreating it from scratch I guess)

**Comment by anyome** [comment]
A2S_Info doesn't works anymore anyone can help please?

**Comment by anyome** [comment]
> > Did they ever announce that enabling the new protocol is safe?
> 
> Not that I am aware of. Apparently this behavior was force enabled within Steamworks API.
> 
> > Can you confirm if Garry's Mod still uses Fortran's RAN1 pseudorandom number generator for challenge generation?
> 
> No as the entire system is inside Steamworks binaries, we have no control over it. (besides just recreating it from scratch I guess)

A2s_Info queries outputs this for my servers: https://i.imgur.com/ryfuVbc.png

But it does works for some other servers I don't know how they did it ! : https://i.imgur.com/H7OVM3k.png

**Comment by jorjic** [comment]
> Apparently this behavior was force enabled within Steamworks API.

Ouch. Hopefully this gets fixed soon. We really don't need a bunch of servers continuing to participate in reflected DDoS because they opted for legacy behavior.

**Comment by Fyu-cedi** [comment]
So we can't do something for that ? 
Cause goes from 130 players to 70 is pretty hard to see in view of all the work we did

**Comment by Kefta** [comment]
> So we can't do something for that ? Cause goes from 130 players to 70 is pretty hard to see in view of all the work we did

> Tell those users to update their Steam clients. Contact your server provider for any issues with firewall rules.

See this official post from Valve:
https://steamcommunity.com/discussions/forum/14/2974028351344359625/

**Comment by michak89** [comment]
> 
> > Tell those users to update their Steam clients. Contact your server provider for any issues with firewall rules.
> 


> 
> My steam client is up to date and nothing has changed about our firewall rules since 2020.

Only today I got 8 reports from players about broken server visibility. Everyone has steam client up to date.

**Comment by jorjic** [comment]
@michak89 Provide your server IP address here and start surveying users who can and can't see it. Ask them to give you a screenshot of their Steam client version. Ask them to give you their IP address and ask them to `tracert` to your server. Collect a packet capture on your server and ask them to try to look for the server while you're capturing packets.

We're not going to find out what's going on by crossing our fingers.

**Comment by michak89** [comment]
It's not connectivity nor ping or ranking issue, anyone can connect via console but sometimes query returns no information - 
Yesterday i saw in the favorites tab how query did not return anything about the server for like few minutes and server was fully operational. 

Unfortunately i am not able to check it out from users' perspective, people are not willing to install wireshark or anything on their computers by some random guy from steam.

For now i try to figure out if it's not something with ovh firewall. if only i could reproduce the problem on myself :/

**Comment by jorjic** [comment]
> Unfortunately i am not able to check it out from users' perspective

You don't need them to give you their packet capture. They're probably having their packets delivered to your server just fine, so the packet capture on your end is just as good.

Again, post your server IP address.

**Comment by michak89** [comment]
Serverside packet capture is great idea, i will do it tomorrow, now testing if problem will occur on x64 srcds. 
We run small hosting business and almost all of our clients have noticed a significant server visibility problems.

Ip address of my private sandbox that is also affected with visibility problems since last update:
193.70.126.91:27015

Normaly we use caching proxy for query and some binary modules that may affect it's behavior, but all of that is now disabled.

**Comment by jorjic** [comment]
Steam says your server is not responding. Here's what's going on:
* Client sends A2S_INFO packet to server without challenge
* Server responds with S2C_CHALLENGE
* Client re-sends A2S_INFO with the challenge number appended at the end

The server sends nothing further.

This was tested with the standard Steam client server browser. I didn't provide a port number, and the traffic capture shows the client making *14 simultaneous* no-challenge A2S_INFO requests to the IP address with different probable source ports (26900, 26901, ... 27015, 27016, 27017). Trying again five minutes later and explicitly providing the port did not send 14 simultaneous packets, but still the server doesn't respond after the first S2C_CHALLENGE.

I'm running Windows 10 with the latest updates installed. Here's my Steam client information, provided by going to the Steam client and clicking 'Help' and then 'About Steam' at the top:

```
Steam client application
Built: Dec 16, 2021, at 14:41:19
Steam API: v020
Steam package versions: 1639697812
```

**Comment by michak89** [comment]
I cannot reproduce this behavior for myself:

![image](https://user-images.githubusercontent.com/40716529/146858408-75f97ff8-39f0-4e0e-be41-b4aba20d8742.png)
![image](https://user-images.githubusercontent.com/40716529/146858421-1710b0a9-8e63-4190-b51d-d01eac3ebc2d.png)
![image](https://user-images.githubusercontent.com/40716529/146858429-aade2eda-dbb4-4c73-af7f-9e0724dcbcae.png)
![image](https://user-images.githubusercontent.com/40716529/146858444-1eaad6de-d50a-4c0f-8661-9a780277a335.png)

my steam browser displays server correctly and chellage works, but there are players who experience this problem in a way You described. 

Steam client application
Built: Dec 16, 2021, at 14:41:19
Steam API: v020
Steam package versions: 1639697812 - same.

**Comment by jorjic** [comment]
I opened up a local server in a docker container and I'm able to query A2S_INFO fine as well. The server requires a challenge number, gets one, and responds happily. The docker image installs the latest version of the server, so I'm completely up to date here.

**Comment by michak89** [comment]
Im testing it over internet, there is my local ip on screenshots because of wireshark's display.

**Comment by jorjic** [comment]
I was going to suggest maybe it was some upstream firewall issue, but I don't think so anymore. I'm able to query your server now. It's still going through the challenge process, only now I'm finally getting a response. Were you connected to the server before? Otherwise it could potentially be a server hibernation problem, but that seems unlikely. I can sit here continuously querying your server and getting responses, so it doesn't seem to be some internal rate-limiting problem.

**Comment by michak89** [comment]
This server has hibernation turned off by sv_hibernate_think 1
I dont know if this cvar has something to do with query, but lua is running there even if its empty.

**Comment by jorjic** [comment]
> Im testing it over internet, there is my local ip on screenshots because of wireshark's display.

Sorry, I completely misread your images.

> I dont know if this cvar has something to do with query

I don't think it does, so that can be ignored too.

It would be useful to confirm with a packet capture on the server end if the challenge response from the client is being received. That would rule out firewall policies.

**Comment by michak89** [comment]
I have confirmed that this problem concerns ovh "antiddos game" service and not the srcds itself.

Over last 2 days i managed to catch it few times, it always looks exactly like for @jorjic, respond to S2C_CHALLENGE never reaches the server even though it was sent from client. I recorded packets on both ends for 2 players affected:

https://drive.google.com/file/d/1FMPypaHwJ2OEHcGxc-VYDHwNjvqJBZcI/view?usp=sharing

Additionally, disabled ovh antyddos game for one server and it works propertly now:

https://i.imgur.com/o0gYK5Q.png
https://i.imgur.com/KLWCpP4.png

I suppose this thread can now be closed, servers hosted outside ovh are never affected, everything seems clear now.
Thanks for help @jorjic

**Comment by jorjic** [comment]
@michak89 That's great. Since you have an OVH account, can you open a support ticket under your account and let them know their application-layer DDoS prevention is interfering with the latest DDoS mitigation update to the Source query protocol?

**Comment by michak89** [comment]
Ovh support was already notified about this but I'm not sure if they took it seriously.

**Comment by jorjic** [comment]
> Ovh support was already notified about this but I'm not sure if they took it seriously.

If you haven't opened a ticket yourself yet, open one and link them to this thread. Provide them with the packet captures demonstrating the problem. Tell them that disabling their anti-DDoS option fixes the problem. Try contacting them [on Twitter.](https://twitter.com/ovh_support_us)

**Comment by michak89** [comment]
They have already received a ticket with appropriate description and link to this github thread. 
I think the more people report this problem, the more likely they will do something.

**Comment by jorjic** [comment]
I don't think they took me seriously, either. Anyone continuing to suffer from this needs to take it up with their service provider directly.

**Comment by Hyro** [comment]
![ffb95bf3](https://user-images.githubusercontent.com/8274697/149630647-4bc39c25-eaec-4b7b-b542-1ad1dfac87a2.png)

**Comment by TuXEKRO** [comment]
Hi, exactly same problem here. In my case with Kimsufi, so I am not able to disable the firewall.

Disabling the server's internal firewall helps the server appear to 50% of users. But there are still some who do not receive it.
This Discord addon also gives me an error because the package is being blocked by OVH / Kimsufi AntiDDoS
```
Discord -> Getting auth token...
Discord -> Failed to authenticate using A2S_RULES, fallbacking to hostname validation...
Discord Error -> Failed getting auth token with the reason: Request timed out to your server 164.132.201.232:27015. Make sure that your firewall is open, server is joinable and visible from the server list.
```

I've opened a support ticket with Kimsufi, but seeing as this thread is a bit old I'm afraid it won't be resolved quickly.

**Comment by TuXEKRO** [comment]
Sorry @robotboy655 for reopening the ticket, but I found a solution.
In Kimsufi (and I imagine that in the rest of OVH services) the traffic goes through an internal firewall and for port 27015 it has special rules to avoid replication UDP attacks (an old attack that was carried out frequently and was a joke of head).

Luckily, when I opened the ticket, they gave me access to the server API, so I was able to eliminate said rule without having to do without the AntiDDoS GAME.

**Summary: Remove from the OVH server API a specific rule for port 27015 named 'hl2'. This solved my problem and also keeps the AntiDDoS GAME with UDP packet protection.**
